RecordInput mini batches for dividing processing among multiple devices.

ahundt opened the pull request:
Adds one component of the [high performance models tensorflow benchmark](https://www.tensorflow.org/performance/performance_models) upstream so it is more readily available. This is based on the minibatch function in [preprocessing.py](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/preprocessing.py) in the example benchmark.

I need to write a test, but wasn't sure if this is the right place to add this functionality to the API.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

mrry commented:
Reassigning to @ekelsen, who added this code and is best placed to rule on API changes!

ekelsen commented on tensorflow/python/ops/data_flow_ops.py:
```diff
@@ -2088,6 +2088,7 @@ class RecordInput(object):
   def __init__(self,
                file_pattern,
                batch_size=1,
+               minibatches=None,
```
I think it's best to add new optional parameters to the end of the list, otherwise you break callers not using named parameters.

ahundt replied:
sure, can do

ekelsen commented on tensorflow/python/ops/data_flow_ops.py:
```diff
@@ -2099,6 +2100,10 @@ def __init__(self,
       file_pattern: File path to the dataset, possibly containing wildcards.
         All matching files will be iterated over each epoch.
       batch_size: How many records to return at a time.
+      minibatches: None by default, otherwise specifies
```
I personally find it more natural to think of minibatches as a multiplier of batch_size, rather than a divisor.  batch_size specifies the size of each mini_batch and minibatches tell you how many you get. 

So if you want 4 batches of 32, (batch_size = 32 and minibatches = 4) rather than (batch_size = 128 and minibatches=4).

Thoughts?

ahundt replied:
I was debating between the two and just picked one, I was thinking you choose a batch size because that is what will fit on your machine(s), then subdivide it.

ahundt replied:
shouldn't it be called a macrobatch if it multiplies? :-)

ekelsen replied:
There could still be a larger concept of batch size assuming there are multiple machines involved, which is why I think the least ambiguous option is to use batch size to describe either the largest or smallest grouping.

Perhaps we can slip the issue by avoiding ML terminology all together, since really we're just trying to ask for N groups of size K.

records_per_group
num_groups

ahundt commented:
@ekelsen tried incorporating some of those changes you suggested, guess there are also some conflicts to resolve

ekelsen commented on tensorflow/python/ops/data_flow_ops.py:
```diff
@@ -2108,12 +2109,17 @@ def __init__(self,
+      num_groups: None by default, otherwise specifies
+        how many groups of records of size records_per_group to create.
+        An example use case is to split processing between devices on one computer.
+        _Must_ be a multiple of batch_size.
```
last line is out of date

ekelsen commented on tensorflow/python/ops/data_flow_ops.py:
```diff
@@ -2108,12 +2109,17 @@ def __init__(self,
         records.
       name: Optional name for the operation.
+      num_groups: None by default, otherwise specifies
+        how many groups of records of size records_per_group to create.
```
"how many groups of size records_per_group to create." seems more clear to me.

ekelsen reviewed (approved):
LGTM, thanks for bearing with me caring too much about the names.

ekelsen reviewed (changes requested):
Approved too soon, tests still need to be added.

ahundt committed:
data_flow_ops.py RecordInput multiple batches for dividing processing among devices.
(+6 squashed commits)
Squashed commits:
[f0a251e] data_flow_ops.py RecordInput python 2 & 3 compatibility
[efcc547] data_flow_ops.py op namespace fix
[41e060c] data_flow_ops.py and record_input_test.py style + syntax fix
[faacef7] data_flow_ops.py typo fix
[eab2446] record_input_test.py test covers RecordInput split batches
[ab4d4ba] data_flow_ops.py RecordInput multiple batches for dividing processing among devices.

ahundt commented:
@ekelsen where should a test go? There doesn't seem to be a `data_flow_ops_test.py`.

I also realized that changing `batch_size` would break backwards compatibility, so I decided to just name the second parameter `batches`.

ahundt commented:
found it https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/record_input_test.py

ekelsen commented:
sgtm

ahundt commented:
@ekelsen ready for another review

ekelsen commented:
LGTM, thanks!

ahundt commented:
Great! Do we need to @ mention a tensorflower for the final steps?

ekelsen commented:
Jenkins, test this please

gunan commented:
Jenkins, test this please.

ahundt commented:
tests pass locally for me now

ahundt commented:
@ekelsen could you re-run the test?

ekelsen commented:
Jenkins, test this please.

martinwicke commented:
Jenkins, test this please.

martinwicke commented:
Broken Mac build rebuilt here: http://ci.tensorflow.org/job/tensorflow-pull-requests-mac/5319/console

ahundt commented:
So you're saying the test passed? My local test was actually on OS X

martinwicke commented:
Yes, the test did pass.

ahundt commented:
Awesome my first real code merged in TF :-) thanks!

ahundt commented:
Oh wait I guess some code took the back door earlier via keras, but still good haha.

martinwicke commented:
Thanks!​

ahundt commented on tensorflow/python/ops/data_flow_ops.py:
```diff
@@ -2183,12 +2190,26 @@ def __init__(self,
+        records = [array_ops.reshape(record, []) for record in records]
+        for index, protobuf in zip(six.moves.range(len(records)), records):
+          batch_index = index % self._batches
+          batch_list[batch_index].append(protobuf)
```
@ekelsen I just realized I forgot to include the equivalent of the lines that [parallel_stack batches](https://github.com/tensorflow/benchmarks/blob/b9221114d1261ec28f103455178abcfbeeb9b498/scripts/tf_cnn_benchmarks/preprocessing.py#L378) together, do you think that belongs here as well or should it stay in user code? 

Leaving things as they are may make sense in this case because this code contains full records they'll need to reorganize according to their features.

```
      for device_index in xrange(self.device_count):
        images[device_index] = tf.parallel_stack(images[device_index])
        label_index_batch[device_index] = tf.concat(labels[device_index], 0)
```

ekelsen commented:
I agree that stacking the full records probably doesn't make sense.

Could maybe consider passing an optional function that turns records into something that can be stacked, and then stacking them?

ahundt commented:
Seems like a technically sound strategy, but maybe there is an alternative that is more straightforward for users?

I don't know the internals of `StagingArea` but perhaps feeding full records directly to that, pulling out the result, then returning might be enough so loading will be fast and users can do as they please with the returned results?

Perhaps the easiest thing to do is I could add a comment recommending users just read the [high performance models guide](https://www.tensorflow.org/performance/performance_models) before using. :-)
